runtime/lua/vim/treesitter/query_linter.lua

```diff
@@ -0,0 +1,146 @@
+    lints[buf],
+    { type = error_type, range = error_range, message = error_text, node_text = node_text }
+  )
+end
+
+local function table_contains(predicate, table)
+  for _, elt in pairs(table) do
+    if predicate(elt) then
+      return true
+    end
+  end
+  return false
+end
+
+local function guess_query_lang(buf, _)
+  local filename = vim.api.nvim_buf_get_name(buf)
+  if filename ~= '' then
+    local ok, query_lang = pcall(vim.fn.fnamemodify, filename, ':p:h:t')
+    return ok and query_lang
+  end
+end
+
+local function normalize_opts(buf, opts)
+  opts = opts or {}
+  opts.langs = opts.langs or guess_query_lang
+  if type(opts.langs) == 'function' then
+    opts.langs = opts.langs(buf, opts)
+  end
+  if type(opts.langs) ~= 'table' then
+    opts.langs = { opts.langs }
+  end
+  opts.lints = vim.tbl_extend('keep', opts.lints or {}, {
+    top_level_query = true,
+    fields = true,
+    nodes = true,
+    syntax_error = true,
+  })
+  return opts
+end
+
+local lint_query = [[
+  (program [(named_node) (list) (grouping)] @toplevel-query)
+  (named_node
+    name: _ @named_node)
+  (anonymous_node
+    name: _ @anonymous_node)
+  (field_definition
+    name: (identifier) @field)
+  (ERROR) @error
+]]
+
+function M.lint(buf, opts)
+  if buf == 0 then
+    buf = vim.api.nvim_get_current_buf()
+  end
+  M.clear(buf)
+  lints[buf] = {}
+  local query = vim.treesitter.query.parse('query', lint_query)
+
+  opts = normalize_opts(buf, opts)
+
+  for _, lang in ipairs(opts.langs or { nil }) do
+    local ok, parser_info = pcall(vim.treesitter.language.inspect, lang)
+    if not ok then
+      parser_info = nil
+    end
+
+    vim.treesitter.get_parser(buf):for_each_tree(function(tree, ltree)
+      if ltree:lang() == 'query' then
+        for id, node, _ in query:iter_captures(tree:root(), buf, 0, -1) do
+          local cap_id = query.captures[id]
+          if cap_id == 'error' and opts.lints.syntax_error then
+            add_lint_for_node(node, buf, 'Syntax Error')
+          end
+
+          if lang and parser_info then
+            if cap_id == 'toplevel-query' and opts.lints.top_level_query then
+              local query_text = vim.treesitter.get_node_text(node, buf)
+              local err
+              ok, err = pcall(vim.treesitter.query.parse, lang, query_text)
+              if not ok then
+                add_lint_for_node(node, buf, 'Invalid Query', err)
+              end
+            elseif (cap_id == 'named_node' or cap_id == 'anonymous_node') and opts.lints.nodes then
+              local node_type = vim.treesitter.get_node_text(node, buf)
+              local is_named = cap_id == 'named_node'
+
+              if not is_named then
+                node_type = node_type:gsub('"(.*)".*$', '%1'):gsub('\\(.)', '%1')
+              end
+
+              local found = vim.tbl_contains(MAGIC_NODE_NAMES, node_type)
+                or table_contains(function(t)
+                  return node_type == t[1] and is_named == t[2]
```

Comment:
But can't you just replace this by `vim.tbl_contains(parser_info.symbols, { node_type, is_named })`?